substratevm/src/com.oracle.svm.agent.core/src/com/oracle/svm/agent/core/AgentBase.java

```diff
@@ -0,0 +1,212 @@
+import static org.graalvm.word.WordFactory.nullPointer;
+
+import com.oracle.svm.core.annotate.AutomaticFeature;
+import com.oracle.svm.core.util.VMError;
+import com.oracle.svm.hosted.FeatureImpl;
+import com.oracle.svm.jni.nativeapi.JNIErrors;
+import com.oracle.svm.jni.nativeapi.JNIVersion;
+import org.graalvm.compiler.api.replacements.Fold;
+import org.graalvm.nativeimage.CurrentIsolate;
+import org.graalvm.nativeimage.ImageSingletons;
+import org.graalvm.nativeimage.Platform;
+import org.graalvm.nativeimage.Platforms;
+import org.graalvm.nativeimage.StackValue;
+import org.graalvm.nativeimage.UnmanagedMemory;
+import org.graalvm.nativeimage.c.function.CEntryPoint;
+import org.graalvm.nativeimage.c.function.CEntryPointLiteral;
+import org.graalvm.nativeimage.c.function.CFunctionPointer;
+import org.graalvm.nativeimage.c.struct.SizeOf;
+import org.graalvm.nativeimage.c.type.CCharPointer;
+import org.graalvm.nativeimage.c.type.WordPointer;
+import org.graalvm.nativeimage.hosted.Feature;
+import org.graalvm.word.PointerBase;
+
+import com.oracle.svm.agent.core.jvmti.JvmtiEnv;
+import com.oracle.svm.agent.core.jvmti.JvmtiEventCallbacks;
+import com.oracle.svm.core.c.function.CEntryPointOptions;
+import com.oracle.svm.core.c.function.CEntryPointSetup;
+import com.oracle.svm.jni.nativeapi.JNIEnvironment;
+import com.oracle.svm.jni.nativeapi.JNIJavaVM;
+import com.oracle.svm.jni.nativeapi.JNIObjectHandle;
+
+import java.util.List;
+
+public abstract class AgentBase {
+
+    private static final CEntryPointLiteral<CFunctionPointer> onVMInitLiteral = CEntryPointLiteral.create(AgentBase.class, "onVMInit", JvmtiEnv.class, JNIEnvironment.class, JNIObjectHandle.class);
+    private static final CEntryPointLiteral<CFunctionPointer> onVMStartLiteral = CEntryPointLiteral.create(AgentBase.class, "onVMStart", JvmtiEnv.class, JNIEnvironment.class);
+    private static final CEntryPointLiteral<CFunctionPointer> onVMDeathLiteral = CEntryPointLiteral.create(AgentBase.class, "onVMDeath", JvmtiEnv.class, JNIEnvironment.class);
+    private static final CEntryPointLiteral<CFunctionPointer> onThreadEndLiteral = CEntryPointLiteral.create(AgentBase.class, "onThreadEnd", JvmtiEnv.class, JNIEnvironment.class,
+                    JNIObjectHandle.class);
+
+    protected abstract JavaHandles constructJavaHandles(JNIEnvironment env);
+
+    protected abstract int onLoadCallback(JNIJavaVM vm, JvmtiEnv jvmti, JvmtiEventCallbacks callbacks, String options);
+
+    protected abstract int onUnloadCallback(JNIJavaVM vm);
+
+    protected abstract void onVMStartCallback(JvmtiEnv jvmti, JNIEnvironment jni);
+
+    protected abstract void onVMInitCallback(JvmtiEnv jvmti, JNIEnvironment jni, @SuppressWarnings("unused") JNIObjectHandle thread);
+
+    protected abstract void onVmDeathCallback(JvmtiEnv jvmti, @SuppressWarnings("unused") JNIEnvironment jni);
+
+    protected abstract int getRequiredJVMTIVersion();
+
+    protected static JavaHandles handles;
```

Comment:
You can make this field non-static and access it via `singleton()`. In order to avoid casts, you could introduce a type parameter `<H extends JavaHandles>` that subclasses can specify for the field type and the return type of `constructJavaHandles`, but that might be overkill.

Reply:
I agree with this, however, I think we can't get away from having a static method in each agent subclass. Right now, `NativeImageAgent` has a static `handles()` method that does the type cast. For this approach, we would either have to re-implement singleton in the subclass with the appropriate cast, or keep a static handles method. If there are no other ways, I would go with reimplementing the `singleton()` method.